src/connector/src/sink/snowflake_connector.rs

```diff
@@ -190,56 +186,59 @@ impl SnowflakeHttpClient {
 pub struct SnowflakeS3Client {
     s3_bucket: String,
     s3_path: Option<String>,
-    s3_client: S3Client,
+    opendal_s3_engine: OpendalObjectStore,
 }
 
 impl SnowflakeS3Client {
-    pub async fn new(
+    pub fn new(
         s3_bucket: String,
         s3_path: Option<String>,
         aws_access_key_id: String,
         aws_secret_access_key: String,
         aws_region: String,
     ) -> Self {
-        let credentials = Credentials::new(
-            aws_access_key_id,
-            aws_secret_access_key,
-            // we don't allow temporary credentials
-            None,
-            None,
-            "rw_sink_to_s3_credentials",
-        );
-
-        let region = RegionProviderChain::first_try(Region::new(aws_region)).or_default_provider();
-
-        let config = aws_config::from_env()
-            .credentials_provider(credentials)
-            .region(region)
-            .load()
-            .await;
-
-        // create the brand new s3 client used to sink files to s3
-        let s3_client = S3Client::new(&config);
+        // just use default configuration here for opendal s3 engine
+        let config = ObjectStoreConfig::default();
+
+        // create the s3 engine for streaming upload to the intermediate s3 bucket
+        // note: this will lead to an internal panic if any credential / intermediate creation
+        // process has error, which may not be acceptable...
+        // but it's hard to gracefully handle the error without modifying downstream return type(s)...
+        let opendal_s3_engine = OpendalObjectStore::new_s3_engine_with_credentials(
+            &s3_bucket,
+            config,
+            &aws_access_key_id,
+            &aws_secret_access_key,
+            &aws_region,
+        )
+        .unwrap();
 
         Self {
             s3_bucket,
             s3_path,
-            s3_client,
+            opendal_s3_engine,
         }
     }
 
     pub async fn sink_to_s3(&self, data: Bytes, file_suffix: String) -> Result<()> {
-        self.s3_client
-            .put_object()
-            .bucket(self.s3_bucket.clone())
-            .key(generate_s3_file_name(self.s3_path.clone(), file_suffix))
-            .body(ByteStream::from(data))
-            .send()
+        let path = generate_s3_file_name(self.s3_path.clone(), file_suffix);
+        let mut uploader = self
+            .opendal_s3_engine
+            .streaming_upload(&path)
             .await
             .map_err(|err| {
-                SinkError::Snowflake(format!("failed to sink data to S3, error: {}", err))
+                SinkError::Snowflake(format!(
+                    "failed to create the streaming uploader of opendal s3 engine, error: {}",
+                    err
+                ))
             })?;
-
+        uploader.write_bytes(data).await.map_err(|err| SinkError::Snowflake(format!("failed to write bytes when streaming uploading to s3 for snowflake sink, error: {}", err)))?;
```

Comment:
Actually the idea of using streaming upload is not to simply change the way to upload a complete `Vec<u8>` buffer, but to replace the buffer with the streaming uploader. The ideal implementation will be,
1. at the very beginning, we create a streaming uploader.
2. when new data comes, we do some conversion and write the data to the streaming uploader
3. when checkpoint barrier comes, we close the writer and commit the file to snowflake
4. when the first next data comes, initialize a new streaming uploader